panic: reflect: reflect.Value.Set using unaddressable value merging nested structures

Hi,

I want to do a deep merge where non-empty dst attributes are overwritten with non-empty src attributes values but slices are appended, is this possible?

Can someone help me understand why this does't work?
```
package main

import (
        "fmt"

        "github.com/imdario/mergo"
)

type My struct {
        Name  string
        Data  []int
        Other map[string]Other
}

type Other struct {
        Foo string
        Bar int
        Baz []string
        Faz map[string]string
}

func main() {
        dst := My{
                Data: []int{1, 2, 3},
                Name: "Dest",
                Other: map[string]Other{
                        "stuff": Other{
                                Foo: "bar",
                                Bar: 0,
                        },
                },
        }

        new := My{
                Data: []int{4, 5},
                Name: "Source",
                Other: map[string]Other{
                        "stuff": Other{
                                Foo: "bar",
                                Bar: 1,
                        },
                        "Morestuff": Other{
                                Foo: "foo",
                                Bar: 2,
                        },
                },
        }

        new2 := My{
                Name: "Source2",
                Other: map[string]Other{
                        "stuff": Other{
                                Foo: "bar",
                                Bar: 1,
                        },
                        "Morestuff": Other{
                                Foo: "foo",
                                Bar: 2,
                                Baz: []string{"something", "wrong"},
                                Faz: map[string]string{
                                        "some": "some2",
                                },
                        },
                },
        }

        fmt.Println(dst)
        fmt.Println(new)
        fmt.Println(new2)
        mergo.Merge(&dst, &new, mergo.WithAppendSlice, mergo.WithOverride)
        fmt.Println(dst)
        mergo.Merge(&dst, &new2, mergo.WithAppendSlice, mergo.WithOverride)
        fmt.Println(dst)
}
```

I get:
```
$ go run test.go
{Dest [1 2 3] map[stuff:{bar 0 [] map[]}]}
{Source [4 5] map[Morestuff:{foo 2 [] map[]} stuff:{bar 1 [] map[]}]}
{Source2 [] map[stuff:{bar 1 [] map[]} Morestuff:{foo 2 [something wrong] map[some:some2]}]}
{Source [1 2 3 4 5] map[Morestuff:{foo 2 [] map[]} stuff:{bar 1 [] map[]}]}
panic: reflect: reflect.Value.Set using unaddressable value

goroutine 1 [running]:
reflect.flag.mustBeAssignable(0x95)
        /usr/local/Cellar/go/1.11.1/libexec/src/reflect/value.go:234 +0x157
reflect.Value.Set(0x10b0760, 0xc000022470, 0x95, 0x10b0760, 0xc000074480, 0x15)
        /usr/local/Cellar/go/1.11.1/libexec/src/reflect/value.go:1397 +0x2f
github.com/imdario/mergo.deepMerge(0x10b0760, 0xc000022470, 0x95, 0x10b0760, 0xc000022430, 0x95, 0xc000095b08, 0x3, 0xc00000a1c0, 0x0, ...)
        /Users/ldavim/go/src/github.com/imdario/mergo/merge.go:83 +0x1388
github.com/imdario/mergo.deepMerge(0x10bcde0, 0xc000022440, 0x99, 0x10bcde0, 0xc000022400, 0x99, 0xc000095b08, 0x2, 0xc00000a1c0, 0x0, ...)
        /Users/ldavim/go/src/github.com/imdario/mergo/merge.go:72 +0x23a7
github.com/imdario/mergo.deepMerge(0x10b0700, 0xc0000741a8, 0x195, 0x10b0700, 0xc000074268, 0x195, 0xc000095b08, 0x1, 0xc00000a1c0, 0x0, ...)
        /Users/ldavim/go/src/github.com/imdario/mergo/merge.go:115 +0x986
github.com/imdario/mergo.deepMerge(0x10b9ba0, 0xc000074180, 0x199, 0x10b9ba0, 0xc000074240, 0x199, 0xc000095b08, 0x0, 0xc00000a1c0, 0x199, ...)
        /Users/ldavim/go/src/github.com/imdario/mergo/merge.go:72 +0x23a7
github.com/imdario/mergo.merge(0x10a4780, 0xc000074180, 0x10a4780, 0xc000074240, 0xc000095ce8, 0x2, 0x2, 0x0, 0x0)
        /Users/ldavim/go/src/github.com/imdario/mergo/merge.go:251 +0x29f
github.com/imdario/mergo.Merge(0x10a4780, 0xc000074180, 0x10a4780, 0xc000074240, 0xc000095ce8, 0x2, 0x2, 0x0, 0x0)
        /Users/ldavim/go/src/github.com/imdario/mergo/merge.go:206 +0x71
main.main()
        /Users/ldavim/Desktop/test.go:72 +0xa28
exit status 2
```

So, the first merge works but the second one fails.

Thanks

--- a/issue66_test.go
+++ b/issue66_test.go
@@ -21,7 +21,7 @@ func TestPrivateSlice(t *testing.T) {
 		t.Fatalf("Error during the merge: %v", err)
 	}
 	if len(p1.PublicStrings) != 3 {
-		t.Error("5 elements should be in 'PublicStrings' field")
+		t.Error("3 elements should be in 'PublicStrings' field, when no append")
 	}
 	if len(p1.privateStrings) != 2 {
 		t.Error("2 elements should be in 'privateStrings' field")

--- a/issue90_test.go
+++ b/issue90_test.go
@@ -1,0 +1,55 @@
+package mergo
+
+import (
+	"testing"
+
+	"github.com/magiconair/properties/assert"
+)
+
+func TestMergoSimpleMap(t *testing.T) {
+	dst := map[string]string{"key1": "loosethis", "key2": "keepthis"}
+	src := map[string]string{"key1": "key10"}
+	exp := map[string]string{"key1": "key10", "key2": "keepthis"}
+	Merge(&dst, src, WithAppendSlice, WithOverride)
+	assert.Equal(t, dst, exp)
+}
+
+type CustomStruct struct {
+	SomeMap map[string]string
+}
+
+var testDataStructMap = []struct {
+	name string
+	src  map[string]CustomStruct
+	dst  map[string]CustomStruct
+	exp  map[string]CustomStruct
+}{
+	{name: "Normal",
+		dst: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "loosethis", "key2": "keepthis"}}},
+		src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+		exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10", "key2": "keepthis"}}},
+	},
+	{name: "Init of struct key", dst: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{}}},
+		src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+		exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+	},
+	{name: "Not Init of struct key", dst: map[string]CustomStruct{},
+		src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+		exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+	},
+	{name: "Nil struct key", dst: map[string]CustomStruct{"a": CustomStruct{SomeMap: nil}},
+		src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+		exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}}},
+}
+
+func TestMergoStructMap(t *testing.T) {
+
+	for _, data := range testDataStructMap {
+		dst := data.dst
+		src := data.src
+		exp := data.exp
+
+		Merge(&dst, src, WithAppendSlice, WithOverride)
+		assert.Equal(t, dst, exp)
+	}
+}

--- a/issue90_test.go
+++ b/issue90_test.go
@@ -10,7 +10,8 @@ func TestMergoSimpleMap(t *testing.T) {
 	dst := map[string]string{"key1": "loosethis", "key2": "keepthis"}
 	src := map[string]string{"key1": "key10"}
 	exp := map[string]string{"key1": "key10", "key2": "keepthis"}
-	Merge(&dst, src, WithAppendSlice, WithOverride)
+	err := Merge(&dst, src, WithAppendSlice, WithOverride)
+	assert.Equal(t, err, nil)
 	assert.Equal(t, dst, exp)
 }
 
@@ -49,7 +50,8 @@ func TestMergoStructMap(t *testing.T) {
 		src := data.src
 		exp := data.exp
 
-		Merge(&dst, src, WithAppendSlice, WithOverride)
+		err := Merge(&dst, src, WithAppendSlice, WithOverride)
+		assert.Equal(t, err, nil)
 		assert.Equal(t, dst, exp)
 	}
 }

--- a/issue90_test.go
+++ b/issue90_test.go
@@ -1,19 +1,9 @@
 package mergo
 
 import (
+	"reflect"
 	"testing"
-
-	"github.com/magiconair/properties/assert"
 )
-
-func TestMergoSimpleMap(t *testing.T) {
-	dst := map[string]string{"key1": "loosethis", "key2": "keepthis"}
-	src := map[string]string{"key1": "key10"}
-	exp := map[string]string{"key1": "key10", "key2": "keepthis"}
-	err := Merge(&dst, src, WithAppendSlice, WithOverride)
-	assert.Equal(t, err, nil)
-	assert.Equal(t, dst, exp)
-}
 
 type CustomStruct struct {
 	SomeMap map[string]string
@@ -50,7 +40,11 @@ func TestMergoStructMap(t *testing.T) {
 		exp := data.exp
 
 		err := Merge(&dst, src, WithAppendSlice, WithOverride)
-		assert.Equal(t, err, nil)
-		assert.Equal(t, dst, exp)
+		if err != nil {
+			t.Errorf("mergo error was not nil, %v", err)
+		}
+		if !reflect.DeepEqual(dst, exp) {
+			t.Errorf("Actual: %#v did not match \nExpected: %#v", dst, exp)
+		}
 	}
 }

--- a/issue90_test.go
+++ b/issue90_test.go
@@ -19,33 +19,32 @@ type CustomStruct struct {
 	SomeMap map[string]string
 }
 
-var testDataStructMap = []struct {
-	name string
-	src  map[string]CustomStruct
-	dst  map[string]CustomStruct
-	exp  map[string]CustomStruct
-}{
-	{name: "Normal",
-		dst: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "loosethis", "key2": "keepthis"}}},
-		src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
-		exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10", "key2": "keepthis"}}},
-	},
-	{name: "Init of struct key", dst: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{}}},
-		src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
-		exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
-	},
-	{name: "Not Init of struct key", dst: map[string]CustomStruct{},
-		src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
-		exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
-	},
-	{name: "Nil struct key", dst: map[string]CustomStruct{"a": CustomStruct{SomeMap: nil}},
-		src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
-		exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}}},
-}
-
 func TestMergoStructMap(t *testing.T) {
+	var testData = []struct {
+		name string
+		src  map[string]CustomStruct
+		dst  map[string]CustomStruct
+		exp  map[string]CustomStruct
+	}{
+		{name: "Normal",
+			dst: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "loosethis", "key2": "keepthis"}}},
+			src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+			exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10", "key2": "keepthis"}}},
+		},
+		{name: "Init of struct key", dst: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{}}},
+			src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+			exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+		},
+		{name: "Not Init of struct key", dst: map[string]CustomStruct{},
+			src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+			exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+		},
+		{name: "Nil struct key", dst: map[string]CustomStruct{"a": CustomStruct{SomeMap: nil}},
+			src: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}},
+			exp: map[string]CustomStruct{"a": CustomStruct{SomeMap: map[string]string{"key1": "key10"}}}},
+	}
 
-	for _, data := range testDataStructMap {
+	for _, data := range testData {
 		dst := data.dst
 		src := data.src
 		exp := data.exp

--- a/map.go
+++ b/map.go
@@ -99,11 +99,11 @@ func deepMap(dst, src reflect.Value, visited map[uintptr]*visit, depth int, conf
 				continue
 			}
 			if srcKind == dstKind {
-				if err = deepMerge(dstElement, srcElement, visited, depth+1, config); err != nil {
+				if _, err = deepMerge(dstElement, srcElement, visited, depth+1, config); err != nil {
 					return
 				}
 			} else if dstKind == reflect.Interface && dstElement.Kind() == reflect.Interface {
-				if err = deepMerge(dstElement, srcElement, visited, depth+1, config); err != nil {
+				if _, err = deepMerge(dstElement, srcElement, visited, depth+1, config); err != nil {
 					return
 				}
 			} else if srcKind == reflect.Map {
@@ -157,7 +157,8 @@ func _map(dst, src interface{}, opts ...func(*Config)) error {
 	// To be friction-less, we redirect equal-type arguments
 	// to deepMerge. Only because arguments can be anything.
 	if vSrc.Kind() == vDst.Kind() {
-		return deepMerge(vDst, vSrc, make(map[uintptr]*visit), 0, config)
+		_, err := deepMerge(vDst, vSrc, make(map[uintptr]*visit), 0, config)
+		return err
 	}
 	switch vSrc.Kind() {
 	case reflect.Struct:

--- a/merge.go
+++ b/merge.go
@@ -145,7 +145,6 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 				k := dstElement.Interface()
 				dstElement = reflect.ValueOf(k)
 			}
-			// dstElement.Set(reflect.ValueOf())
 
 			switch srcElement.Kind() {
 			case reflect.Chan, reflect.Func, reflect.Map, reflect.Interface, reflect.Slice:

--- a/merge.go
+++ b/merge.go
@@ -98,17 +98,16 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 			for i, n := 0, dst.NumField(); i < n; i++ {
 				dstField := dst.Field(i)
 				structField := dst.Type().Field(i)
+				// copy un-exported struct fields
 				if !isExportedComponent(&structField) {
 					rf := dstCp.Field(i)
-					rf = reflect.NewAt(rf.Type(), unsafe.Pointer(rf.UnsafeAddr())).Elem() // no:gosec
-
+					rf = reflect.NewAt(rf.Type(), unsafe.Pointer(rf.Pointer())).Elem() //nolint:gosec
 					dstRF := dst.Field(i)
 					if !dst.Field(i).CanAddr() {
 						continue
 					}
 
-					dstRF = reflect.NewAt(dstRF.Type(), unsafe.Pointer(dstRF.UnsafeAddr())).Elem() // no:gosec
-
+					dstRF = reflect.NewAt(dstRF.Type(), unsafe.Pointer(dstRF.UnsafeAddr())).Elem() //nolint:gosec
 					rf.Set(dstRF)
 					continue
 				}

--- a/merge.go
+++ b/merge.go
@@ -180,7 +180,7 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 			}
 			newSlice = src
 		} else if config.AppendSlice {
-			if src.Type() != dst.Type() {
+			if typeCheck && src.Type() != dst.Type() {
 				err = fmt.Errorf("cannot append two slice with different type (%s, %s)", src.Type(), dst.Type())
 				return
 			}

--- a/merge.go
+++ b/merge.go
@@ -174,7 +174,10 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 		}
 	case reflect.Slice:
 		newSlice := dst
-		if (!isEmptyValue(src) || overwriteWithEmptySrc) && (overwrite || isEmptyValue(dst)) && !config.AppendSlice {
+		if (!isEmptyValue(src) || overwriteWithEmptySrc || overwriteSliceWithEmptySrc) && (overwrite || isEmptyValue(dst)) && !config.AppendSlice {
+			if typeCheck && src.Type() != dst.Type() {
+				return fmt.Errorf("cannot override two slices with different type (%s, %s)", src.Type(), dst.Type())
+			}
 			newSlice = src
 		} else if config.AppendSlice {
 			if src.Type() != dst.Type() {

--- a/merge.go
+++ b/merge.go
@@ -146,7 +146,7 @@ func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, co
 					dst.SetMapIndex(key, dstSlice)
 				}
 			}
-			if dstElement.IsValid() && !isEmptyValue(dstElement) && (reflect.TypeOf(srcElement.Interface()).Kind() == reflect.Map || reflect.TypeOf(srcElement.Interface()).Kind() == reflect.Slice) {
+			if dstElement.IsValid() && !isEmptyValue(dstElement) && (reflect.TypeOf(srcElement.Interface()).Kind() == reflect.Map || reflect.TypeOf(srcElement.Interface()).Kind() == reflect.Slice) || (reflect.TypeOf(srcElement.Interface()).Kind() == reflect.Struct) {
 				continue
 			}
 

--- a/merge.go
+++ b/merge.go
@@ -17,10 +17,8 @@ import (
 func hasExportedField(dst reflect.Value) (exported bool) {
 	for i, n := 0, dst.NumField(); i < n; i++ {
 		field := dst.Type().Field(i)
-		if field.Anonymous && dst.Field(i).Kind() == reflect.Struct {
-			exported = exported || hasExportedField(dst.Field(i))
-		} else {
-			exported = exported || isFieldExported(field)
+		if isFieldExported(field) {
+			return true
 		}
 	}
 	return
@@ -68,6 +66,7 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 	if !src.IsValid() {
 		return
 	}
+
 	if dst.CanAddr() {
 		addr := dst.UnsafeAddr()
 		h := 17 * addr
@@ -87,6 +86,11 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 			err = fn(dst, src)
 			return
 		}
+	}
+
+	if dst.IsValid() && src.IsValid() && src.Type() != dst.Type() {
+		err = fmt.Errorf("cannot append two different types (%s, %s)", src.Kind(), dst.Kind())
+		return
 	}
 
 	switch dst.Kind() {
@@ -121,11 +125,13 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 			} else {
 				dst = dstCp
 			}
+			return
 		} else {
 			if (isReflectNil(dst) || overwrite) && (!isEmptyValue(src) || overwriteWithEmptySrc) {
 				dst = src
 			}
 		}
+
 	case reflect.Map:
 		if dst.IsNil() && !src.IsNil() {
 			if dst.CanSet() {
@@ -137,91 +143,52 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 		}
 		for _, key := range src.MapKeys() {
 			srcElement := src.MapIndex(key)
+			dstElement := dst.MapIndex(key)
 			if !srcElement.IsValid() {
 				continue
 			}
-			dstElement := dst.MapIndex(key)
 			if dst.MapIndex(key).IsValid() {
 				k := dstElement.Interface()
 				dstElement = reflect.ValueOf(k)
 			}
-
-			switch srcElement.Kind() {
-			case reflect.Chan, reflect.Func, reflect.Map, reflect.Interface, reflect.Slice:
-				if srcElement.IsNil() {
-					continue
+			if isReflectNil(srcElement) {
+				if overwrite || isReflectNil(dstElement) {
+					dst.SetMapIndex(key, srcElement)
 				}
-				fallthrough
-			default:
-				if !srcElement.CanInterface() {
-					continue
-				}
-				switch reflect.TypeOf(srcElement.Interface()).Kind() {
-				case reflect.Struct:
-					fallthrough
-				case reflect.Ptr:
-					fallthrough
-				case reflect.Map:
-					srcMapElm := srcElement
-					dstMapElm := dstElement
-					if srcMapElm.CanInterface() {
-						srcMapElm = reflect.ValueOf(srcMapElm.Interface())
-						if dstMapElm.IsValid() {
-							dstMapElm = reflect.ValueOf(dstMapElm.Interface())
-						}
-					}
-					dstMapElm, err = deepMerge(dstMapElm, srcMapElm, visited, depth+1, config)
-					if err != nil {
-						return
-					}
-					dst.SetMapIndex(key, dstMapElm)
-				case reflect.Slice:
-					srcSlice := reflect.ValueOf(srcElement.Interface())
-
-					var dstSlice reflect.Value
-					if !dstElement.IsValid() || isReflectNil(dstElement) {
-						dstSlice = reflect.MakeSlice(srcSlice.Type(), 0, srcSlice.Len())
-					} else {
-						dstSlice = reflect.ValueOf(dstElement.Interface())
-					}
-
-					if (!isEmptyValue(src) || overwriteWithEmptySrc || overwriteSliceWithEmptySrc) && (overwrite || isEmptyValue(dst)) && !config.AppendSlice {
-						if typeCheck && srcSlice.Type() != dstSlice.Type() {
-							return fmt.Errorf("cannot override two slices with different type (%s, %s)", srcSlice.Type(), dstSlice.Type())
-						}
-						dstSlice = srcSlice
-					} else if config.AppendSlice {
-						if srcSlice.Type() != dstSlice.Type() {
-							return fmt.Errorf("cannot append two slices with different type (%s, %s)", srcSlice.Type(), dstSlice.Type())
-						}
-						dstSlice = reflect.AppendSlice(dstSlice, srcSlice)
-					}
-					dst.SetMapIndex(key, dstSlice)
-				}
+				continue
 			}
-			if dstElement.IsValid() && !isEmptyValue(dstElement) && (reflect.TypeOf(srcElement.Interface()).Kind() == reflect.Map || reflect.TypeOf(srcElement.Interface()).Kind() == reflect.Slice) || (reflect.TypeOf(srcElement.Interface()).Kind() == reflect.Struct) {
+			if !srcElement.CanInterface() {
 				continue
 			}
 
-			if srcElement.IsValid() && ((srcElement.Kind() != reflect.Ptr && overwrite) || !dstElement.IsValid() || isEmptyValue(dstElement)) {
-				if dst.IsNil() {
-					dst.Set(reflect.MakeMap(dst.Type()))
+			if srcElement.CanInterface() {
+				srcElement = reflect.ValueOf(srcElement.Interface())
+				if dstElement.IsValid() {
+					dstElement = reflect.ValueOf(dstElement.Interface())
 				}
-				dst.SetMapIndex(key, srcElement)
 			}
+			dstElement, err = deepMerge(dstElement, srcElement, visited, depth+1, config)
+			if err != nil {
+				return
+			}
+			dst.SetMapIndex(key, dstElement)
+
 		}
 	case reflect.Slice:
-		if !dst.CanSet() {
-			break
-		}
-		if (!isEmptyValue(src) || overwriteWithEmptySrc || overwriteSliceWithEmptySrc) && (overwrite || isEmptyValue(dst)) && !config.AppendSlice {
-			dst.Set(src)
+		newSlice := dst
+		if (!isEmptyValue(src) || overwriteWithEmptySrc) && (overwrite || isEmptyValue(dst)) && !config.AppendSlice {
+			newSlice = src
 		} else if config.AppendSlice {
 			if src.Type() != dst.Type() {
 				err = fmt.Errorf("cannot append two slice with different type (%s, %s)", src.Type(), dst.Type())
 				return
 			}
-			dst.Set(reflect.AppendSlice(dst, src))
+			newSlice = reflect.AppendSlice(dst, src)
+		}
+		if dst.CanSet() {
+			dst.Set(newSlice)
+		} else {
+			dst = newSlice
 		}
 	case reflect.Ptr, reflect.Interface:
 		if isReflectNil(src) {
@@ -237,7 +204,6 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 						dst = src
 					}
 				}
-<<<<<<< HEAD
 			}
 			break
 		}
@@ -247,9 +213,6 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 				if dst.CanSet() && (overwrite || isEmptyValue(dst)) {
 					dst.Set(src)
 				}
-=======
-
->>>>>>> semi working
 			} else if src.Kind() == reflect.Ptr {
 				if dst, err = deepMerge(dst.Elem(), src.Elem(), visited, depth+1, config); err != nil {
 					return

--- a/merge.go
+++ b/merge.go
@@ -19,10 +19,32 @@ func hasExportedField(dst reflect.Value) (exported bool) {
 		if field.Anonymous && dst.Field(i).Kind() == reflect.Struct {
 			exported = exported || hasExportedField(dst.Field(i))
 		} else {
-			exported = exported || len(field.PkgPath) == 0
+			exported = exported || isFieldExported(field)
 		}
 	}
 	return
+}
+
+func isFieldExported(field reflect.StructField) bool {
+	if len(field.PkgPath) > 0 {
+		return false
+	}
+	c := field.Name[0]
+	if 'a' <= c && c <= 'z' || c == '_' {
+		return false
+	}
+	return true
+}
+
+func isTypeExported(v reflect.Type) bool {
+	if len(v.PkgPath()) > 0 {
+		return false
+	}
+	c := v.Name()[0]
+	if 'a' <= c && c <= 'z' || c == '_' {
+		return false
+	}
+	return true
 }
 
 type Config struct {
@@ -280,4 +302,84 @@ func merge(dst, src interface{}, opts ...func(*Config)) error {
 		return ErrDifferentArgumentsTypes
 	}
 	return deepMerge(vDst, vSrc, make(map[uintptr]*visit), 0, config)
+}
+
+func handleNil(dst reflect.Value) reflect.Value {
+	fmt.Println("handlesNil(1.0)", dst, dst.CanSet(), dst.Kind())
+	fmt.Println("handleNil(1.0.1)", dst, dst.CanSet(), isReflectNil(dst), isEmptyValue(dst))
+	if !dst.CanSet() {
+		t := reflect.Indirect(reflect.ValueOf(dst)).Type()
+		dsttmp := reflect.New(t).Elem()
+		fmt.Println(dsttmp)
+		initiallize(t, dsttmp)
+		fmt.Println("dsttmp", dsttmp)
+		fmt.Println("handleNil(1.2.0.0)", dsttmp, dsttmp.CanSet(), dst, dst.CanSet(), dst.Kind())
+		if !isTypeExported(t) {
+			return dsttmp
+		}
+		dsttmp.Set(dst)
+		dst = dsttmp
+		fmt.Println("handleNil(1.2.0)", dst, dst.CanSet(), dst.Kind())
+	}
+	fmt.Println("handleNil(1.2)", dst, dst.CanSet(), dst.Kind())
+	return dst
+}
+
+func initiallize(t reflect.Type, v reflect.Value) {
+	// fmt.Println("initializeStruct(1.2.1)", t)
+	switch t.Kind() {
+	case reflect.Map:
+		v.Set(reflect.MakeMap(t))
+	case reflect.Slice:
+		v.Set(reflect.MakeSlice(t, 0, 0))
+	case reflect.Chan:
+		v.Set(reflect.MakeChan(t, 0))
+	case reflect.Struct:
+		for i := 0; i < v.NumField(); i++ {
+			f := v.Field(i)
+			ft := t.Field(i)
+			initiallize(ft.Type, f)
+		}
+	case reflect.Ptr:
+		ft := t.Elem()
+		fv := reflect.New(ft)
+		initiallize(ft, fv.Elem())
+		if isTypeExported(ft) {
+			v.Set(fv)
+		}
+	default:
+	}
+}
+
+func handleEmpty(dst, src reflect.Value, overwrite, appendSlice, overwriteWithEmptySrc bool) reflect.Value {
+	fmt.Println("handleEmpty(0)", dst, src, dst.CanSet(), dst.Kind(), src.Kind())
+	dst = handleNil(dst)
+	overwrite = (overwrite && !appendSlice || isEmptyValue(dst))
+	fmt.Println("handleEmpty(3.0)", dst, src, dst.CanSet(), dst.Kind(), src.Kind(), "can set: ", dst.CanSet(), "!isEmptyValue: ", !isEmptyValue(src), overwriteWithEmptySrc, overwrite)
+	if (!isEmptyValue(src) || overwriteWithEmptySrc) && overwrite {
+		fmt.Println("handleEmpty(3.1)", dst, src)
+		if dst.CanSet() {
+			dst.Set(src)
+		} else {
+			dst = src
+		}
+	}
+	fmt.Println("handleEmpty(4)", dst, src, dst.CanSet())
+	if dst.CanInterface() {
+		dst = reflect.ValueOf(dst.Interface())
+	}
+	return dst
+}
+
+// IsReflectNil is the reflect value provided nil
+func isReflectNil(v reflect.Value) bool {
+	k := v.Kind()
+	switch k {
+	case reflect.Interface, reflect.Slice, reflect.Chan, reflect.Func, reflect.Map, reflect.Ptr:
+		// Both interface and slice are nil if first word is 0.
+		// Both are always bigger than a word; assume flagIndir.
+		return v.IsNil()
+	default:
+		return false
+	}
 }

--- a/merge.go
+++ b/merge.go
@@ -26,21 +26,18 @@ func hasExportedField(dst reflect.Value) (exported bool) {
 }
 
 func isFieldExported(field reflect.StructField) bool {
-	if len(field.PkgPath) > 0 {
-		return false
-	}
-	c := field.Name[0]
-	if 'a' <= c && c <= 'z' || c == '_' {
-		return false
-	}
-	return true
+	return isExportedComponent(field.Name, field.PkgPath)
 }
 
 func isTypeExported(v reflect.Type) bool {
-	if len(v.PkgPath()) > 0 {
+	return isExportedComponent(v.Name(), v.PkgPath())
+}
+
+func isExportedComponent(name, pkgPath string) bool {
+	if len(pkgPath) > 0 {
 		return false
 	}
-	c := v.Name()[0]
+	c := name[0]
 	if 'a' <= c && c <= 'z' || c == '_' {
 		return false
 	}
@@ -63,7 +60,8 @@ type Transformers interface {
 // Traverses recursively both values, assigning src's fields values to dst.
 // The map argument tracks comparisons that have already been seen, which allows
 // short circuiting on recursive types.
-func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, config *Config) (err error) {
+func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, config *Config) (dst reflect.Value, err error) {
+	dst = dstIn
 	overwrite := config.Overwrite
 	typeCheck := config.TypeCheck
 	overwriteWithEmptySrc := config.overwriteWithEmptyValue
@@ -80,7 +78,7 @@ func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, co
 		typ := dst.Type()
 		for p := seen; p != nil; p = p.next {
 			if p.ptr == addr && p.typ == typ {
-				return nil
+				return dst, nil
 			}
 		}
 		// Remember, remember...
@@ -97,11 +95,28 @@ func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, co
 	switch dst.Kind() {
 	case reflect.Struct:
 		if hasExportedField(dst) {
+			dstCp := reflect.New(dst.Type()).Elem()
 			for i, n := 0, dst.NumField(); i < n; i++ {
-				if err = deepMerge(dst.Field(i), src.Field(i), visited, depth+1, config); err != nil {
+				dstField := dst.Field(i)
+				if !isFieldExported(dst.Type().Field(i)) {
+					continue
+				}
+				if dst.Field(i).IsValid() {
+					k := dstField.Interface()
+					dstField = reflect.ValueOf(k)
+				}
+				dstField, err = deepMerge(dstField, src.Field(i), visited, depth+1, config)
+				if err != nil {
 					return
 				}
+				dstCp.Field(i).Set(dstField)
 			}
+			if dst.CanSet() {
+				dst.Set(dstCp)
+			} else {
+				dst = dstCp
+			}
+
 		} else {
 			if dst.CanSet() && (!isEmptyValue(src) || overwriteWithEmptySrc) && (overwrite || isEmptyValue(dst)) {
 				dst.Set(src)
@@ -109,7 +124,12 @@ func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, co
 		}
 	case reflect.Map:
 		if dst.IsNil() && !src.IsNil() {
-			dst.Set(reflect.MakeMap(dst.Type()))
+			if dst.CanSet() {
+				dst.Set(reflect.MakeMap(dst.Type()))
+			} else {
+				dst = src
+				return
+			}
 		}
 		for _, key := range src.MapKeys() {
 			srcElement := src.MapIndex(key)
@@ -117,6 +137,12 @@ func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, co
 				continue
 			}
 			dstElement := dst.MapIndex(key)
+			if dst.MapIndex(key).IsValid() {
+				k := dstElement.Interface()
+				dstElement = reflect.ValueOf(k)
+			}
+			// dstElement.Set(reflect.ValueOf())
+
 			switch srcElement.Kind() {
 			case reflect.Chan, reflect.Func, reflect.Map, reflect.Interface, reflect.Slice:
 				if srcElement.IsNil() {
@@ -141,9 +167,11 @@ func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, co
 							dstMapElm = reflect.ValueOf(dstMapElm.Interface())
 						}
 					}
-					if err = deepMerge(dstMapElm, srcMapElm, visited, depth+1, config); err != nil {
+					dstMapElm, err = deepMerge(dstMapElm, srcMapElm, visited, depth+1, config)
+					if err != nil {
 						return
 					}
+					dst.SetMapIndex(key, dstMapElm)
 				case reflect.Slice:
 					srcSlice := reflect.ValueOf(srcElement.Interface())
 
@@ -187,22 +215,26 @@ func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, co
 			dst.Set(src)
 		} else if config.AppendSlice {
 			if src.Type() != dst.Type() {
-				return fmt.Errorf("cannot append two slice with different type (%s, %s)", src.Type(), dst.Type())
+				err = fmt.Errorf("cannot append two slice with different type (%s, %s)", src.Type(), dst.Type())
+				return
 			}
 			dst.Set(reflect.AppendSlice(dst, src))
 		}
-	case reflect.Ptr:
-		fallthrough
-	case reflect.Interface:
-		if src.IsNil() {
+	case reflect.Ptr, reflect.Interface:
+		if isReflectNil(src) {
 			break
 		}
 
 		if dst.Kind() != reflect.Ptr && src.Type().AssignableTo(dst.Type()) {
 			if dst.IsNil() || overwrite {
-				if dst.CanSet() && (overwrite || isEmptyValue(dst)) {
-					dst.Set(src)
+				if overwrite || isEmptyValue(dst) {
+					if dst.CanSet() {
+						dst.Set(src)
+					} else {
+						dst = src
+					}
 				}
+<<<<<<< HEAD
 			}
 			break
 		}
@@ -212,16 +244,19 @@ func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, co
 				if dst.CanSet() && (overwrite || isEmptyValue(dst)) {
 					dst.Set(src)
 				}
+=======
+
+>>>>>>> semi working
 			} else if src.Kind() == reflect.Ptr {
-				if err = deepMerge(dst.Elem(), src.Elem(), visited, depth+1, config); err != nil {
+				if dst, err = deepMerge(dst.Elem(), src.Elem(), visited, depth+1, config); err != nil {
 					return
 				}
 			} else if dst.Elem().Type() == src.Type() {
-				if err = deepMerge(dst.Elem(), src, visited, depth+1, config); err != nil {
+				if dst, err = deepMerge(dst.Elem(), src, visited, depth+1, config); err != nil {
 					return
 				}
 			} else {
-				return ErrDifferentArgumentsTypes
+				return dst, ErrDifferentArgumentsTypes
 			}
 			break
 		}
@@ -229,12 +264,17 @@ func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, co
 			if dst.CanSet() && (overwrite || isEmptyValue(dst)) {
 				dst.Set(src)
 			}
-		} else if err = deepMerge(dst.Elem(), src.Elem(), visited, depth+1, config); err != nil {
+			// TODO HERE
+		} else if _, err = deepMerge(dst.Elem(), src.Elem(), visited, depth+1, config); err != nil {
 			return
 		}
+		// dst.Set()
 	default:
-		if dst.CanSet() && (!isEmptyValue(src) || overwriteWithEmptySrc) && (overwrite || isEmptyValue(dst)) {
+		overwrite := (!isEmptyValue(src) || overwriteWithEmptySrc) && (overwrite || isEmptyValue(dst))
+		if dst.CanSet() && overwrite {
 			dst.Set(src)
+		} else {
+			dst = src
 		}
 	}
 
@@ -298,30 +338,28 @@ func merge(dst, src interface{}, opts ...func(*Config)) error {
 	if vDst, vSrc, err = resolveValues(dst, src); err != nil {
 		return err
 	}
+	if !vDst.CanSet() {
+		return fmt.Errorf("cannot set dst, needs reference")
+	}
 	if vDst.Type() != vSrc.Type() {
 		return ErrDifferentArgumentsTypes
 	}
-	return deepMerge(vDst, vSrc, make(map[uintptr]*visit), 0, config)
+	_, err = deepMerge(vDst, vSrc, make(map[uintptr]*visit), 0, config)
+	return err
 }
 
 func handleNil(dst reflect.Value) reflect.Value {
-	fmt.Println("handlesNil(1.0)", dst, dst.CanSet(), dst.Kind())
-	fmt.Println("handleNil(1.0.1)", dst, dst.CanSet(), isReflectNil(dst), isEmptyValue(dst))
 	if !dst.CanSet() {
 		t := reflect.Indirect(reflect.ValueOf(dst)).Type()
 		dsttmp := reflect.New(t).Elem()
 		fmt.Println(dsttmp)
 		initiallize(t, dsttmp)
-		fmt.Println("dsttmp", dsttmp)
-		fmt.Println("handleNil(1.2.0.0)", dsttmp, dsttmp.CanSet(), dst, dst.CanSet(), dst.Kind())
 		if !isTypeExported(t) {
 			return dsttmp
 		}
 		dsttmp.Set(dst)
 		dst = dsttmp
-		fmt.Println("handleNil(1.2.0)", dst, dst.CanSet(), dst.Kind())
 	}
-	fmt.Println("handleNil(1.2)", dst, dst.CanSet(), dst.Kind())
 	return dst
 }
 
@@ -352,19 +390,15 @@ func initiallize(t reflect.Type, v reflect.Value) {
 }
 
 func handleEmpty(dst, src reflect.Value, overwrite, appendSlice, overwriteWithEmptySrc bool) reflect.Value {
-	fmt.Println("handleEmpty(0)", dst, src, dst.CanSet(), dst.Kind(), src.Kind())
 	dst = handleNil(dst)
 	overwrite = (overwrite && !appendSlice || isEmptyValue(dst))
-	fmt.Println("handleEmpty(3.0)", dst, src, dst.CanSet(), dst.Kind(), src.Kind(), "can set: ", dst.CanSet(), "!isEmptyValue: ", !isEmptyValue(src), overwriteWithEmptySrc, overwrite)
 	if (!isEmptyValue(src) || overwriteWithEmptySrc) && overwrite {
-		fmt.Println("handleEmpty(3.1)", dst, src)
 		if dst.CanSet() {
 			dst.Set(src)
 		} else {
 			dst = src
 		}
 	}
-	fmt.Println("handleEmpty(4)", dst, src, dst.CanSet())
 	if dst.CanInterface() {
 		dst = reflect.ValueOf(dst.Interface())
 	}

--- a/merge.go
+++ b/merge.go
@@ -17,14 +17,14 @@ import (
 func hasExportedField(dst reflect.Value) (exported bool) {
 	for i, n := 0, dst.NumField(); i < n; i++ {
 		field := dst.Type().Field(i)
-		if isExportedComponent(field) {
+		if isExportedComponent(&field) {
 			return true
 		}
 	}
 	return
 }
 
-func isExportedComponent(field reflect.StructField) bool {
+func isExportedComponent(field *reflect.StructField) bool {
 	name := field.Name
 	pkgPath := field.PkgPath
 	if len(pkgPath) > 0 {
@@ -97,7 +97,8 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 			dstCp := reflect.New(dst.Type()).Elem()
 			for i, n := 0, dst.NumField(); i < n; i++ {
 				dstField := dst.Field(i)
-				if !isExportedComponent(dst.Type().Field(i)) {
+				structField := dst.Type().Field(i)
+				if !isExportedComponent(&structField) {
 					rf := dstCp.Field(i)
 					rf = reflect.NewAt(rf.Type(), unsafe.Pointer(rf.UnsafeAddr())).Elem()
 

--- a/merge.go
+++ b/merge.go
@@ -100,14 +100,14 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 				structField := dst.Type().Field(i)
 				if !isExportedComponent(&structField) {
 					rf := dstCp.Field(i)
-					rf = reflect.NewAt(rf.Type(), unsafe.Pointer(rf.UnsafeAddr())).Elem()
+					rf = reflect.NewAt(rf.Type(), unsafe.Pointer(rf.UnsafeAddr())).Elem() // no:gosec
 
 					dstRF := dst.Field(i)
 					if !dst.Field(i).CanAddr() {
 						continue
 					}
 
-					dstRF = reflect.NewAt(dstRF.Type(), unsafe.Pointer(dstRF.UnsafeAddr())).Elem()
+					dstRF = reflect.NewAt(dstRF.Type(), unsafe.Pointer(dstRF.UnsafeAddr())).Elem() // no:gosec
 
 					rf.Set(dstRF)
 					continue

--- a/merge.go
+++ b/merge.go
@@ -17,18 +17,16 @@ import (
 func hasExportedField(dst reflect.Value) (exported bool) {
 	for i, n := 0, dst.NumField(); i < n; i++ {
 		field := dst.Type().Field(i)
-		if isFieldExported(field) {
+		if isExportedComponent(field) {
 			return true
 		}
 	}
 	return
 }
 
-func isFieldExported(field reflect.StructField) bool {
-	return isExportedComponent(field.Name, field.PkgPath)
-}
-
-func isExportedComponent(name, pkgPath string) bool {
+func isExportedComponent(field reflect.StructField) bool {
+	name := field.Name
+	pkgPath := field.PkgPath
 	if len(pkgPath) > 0 {
 		return false
 	}
@@ -99,7 +97,7 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 			dstCp := reflect.New(dst.Type()).Elem()
 			for i, n := 0, dst.NumField(); i < n; i++ {
 				dstField := dst.Field(i)
-				if !isFieldExported(dst.Type().Field(i)) {
+				if !isExportedComponent(dst.Type().Field(i)) {
 					rf := dstCp.Field(i)
 					rf = reflect.NewAt(rf.Type(), unsafe.Pointer(rf.UnsafeAddr())).Elem()
 

--- a/merge.go
+++ b/merge.go
@@ -180,7 +180,7 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 					srcSlice := reflect.ValueOf(srcElement.Interface())
 
 					var dstSlice reflect.Value
-					if !dstElement.IsValid() || dstElement.IsNil() {
+					if !dstElement.IsValid() || isReflectNil(dstElement) {
 						dstSlice = reflect.MakeSlice(srcSlice.Type(), 0, srcSlice.Len())
 					} else {
 						dstSlice = reflect.ValueOf(dstElement.Interface())

--- a/merge.go
+++ b/merge.go
@@ -11,6 +11,7 @@ package mergo
 import (
 	"fmt"
 	"reflect"
+	"unsafe"
 )
 
 func hasExportedField(dst reflect.Value) (exported bool) {
@@ -99,27 +100,34 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 			for i, n := 0, dst.NumField(); i < n; i++ {
 				dstField := dst.Field(i)
 				if !isFieldExported(dst.Type().Field(i)) {
+					rf := dstCp.Field(i)
+					rf = reflect.NewAt(rf.Type(), unsafe.Pointer(rf.UnsafeAddr())).Elem()
+
+					dstRF := dst.Field(i)
+					if !dst.Field(i).CanAddr() {
+						continue
+					}
+
+					dstRF = reflect.NewAt(dstRF.Type(), unsafe.Pointer(dstRF.UnsafeAddr())).Elem()
+
+					rf.Set(dstRF)
 					continue
 				}
-				if dst.Field(i).IsValid() {
-					k := dstField.Interface()
-					dstField = reflect.ValueOf(k)
-				}
-				dstField, err = deepMerge(dstField, src.Field(i), visited, depth+1, config)
+				dstField, err = deepMerge(dst.Field(i), src.Field(i), visited, depth+1, config)
 				if err != nil {
 					return
 				}
 				dstCp.Field(i).Set(dstField)
 			}
+
 			if dst.CanSet() {
 				dst.Set(dstCp)
 			} else {
 				dst = dstCp
 			}
-
 		} else {
-			if dst.CanSet() && (!isEmptyValue(src) || overwriteWithEmptySrc) && (overwrite || isEmptyValue(dst)) {
-				dst.Set(src)
+			if (isReflectNil(dst) || overwrite) && (!isEmptyValue(src) || overwriteWithEmptySrc) {
+				dst = src
 			}
 		}
 	case reflect.Map:
@@ -251,6 +259,7 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 				if dst, err = deepMerge(dst.Elem(), src.Elem(), visited, depth+1, config); err != nil {
 					return
 				}
+				dst = dst.Addr()
 			} else if dst.Elem().Type() == src.Type() {
 				if dst, err = deepMerge(dst.Elem(), src, visited, depth+1, config); err != nil {
 					return
@@ -261,20 +270,24 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 			break
 		}
 		if dst.IsNil() || overwrite {
-			if dst.CanSet() && (overwrite || isEmptyValue(dst)) {
-				dst.Set(src)
+			if (overwrite || isEmptyValue(dst)) && (overwriteWithEmptySrc || !isEmptyValue(src)) {
+				if dst.CanSet() {
+					dst.Set(src)
+				} else {
+					dst = src
+				}
 			}
-			// TODO HERE
 		} else if _, err = deepMerge(dst.Elem(), src.Elem(), visited, depth+1, config); err != nil {
 			return
 		}
-		// dst.Set()
 	default:
-		overwrite := (!isEmptyValue(src) || overwriteWithEmptySrc) && (overwrite || isEmptyValue(dst))
-		if dst.CanSet() && overwrite {
-			dst.Set(src)
-		} else {
-			dst = src
+		overwriteFull := (!isEmptyValue(src) || overwriteWithEmptySrc) && (overwrite || isEmptyValue(dst))
+		if overwriteFull {
+			if dst.CanSet() {
+				dst.Set(src)
+			} else {
+				dst = src
+			}
 		}
 	}
 

--- a/merge.go
+++ b/merge.go
@@ -30,10 +30,6 @@ func isFieldExported(field reflect.StructField) bool {
 	return isExportedComponent(field.Name, field.PkgPath)
 }
 
-func isTypeExported(v reflect.Type) bool {
-	return isExportedComponent(v.Name(), v.PkgPath())
-}
-
 func isExportedComponent(name, pkgPath string) bool {
 	if len(pkgPath) > 0 {
 		return false
@@ -113,7 +109,7 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 					rf.Set(dstRF)
 					continue
 				}
-				dstField, err = deepMerge(dst.Field(i), src.Field(i), visited, depth+1, config)
+				dstField, err = deepMerge(dstField, src.Field(i), visited, depth+1, config)
 				if err != nil {
 					return
 				}

--- a/merge.go
+++ b/merge.go
@@ -176,7 +176,7 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 		newSlice := dst
 		if (!isEmptyValue(src) || overwriteWithEmptySrc || overwriteSliceWithEmptySrc) && (overwrite || isEmptyValue(dst)) && !config.AppendSlice {
 			if typeCheck && src.Type() != dst.Type() {
-				return fmt.Errorf("cannot override two slices with different type (%s, %s)", src.Type(), dst.Type())
+				return dst, fmt.Errorf("cannot override two slices with different type (%s, %s)", src.Type(), dst.Type())
 			}
 			newSlice = src
 		} else if config.AppendSlice {

--- a/merge.go
+++ b/merge.go
@@ -348,63 +348,6 @@ func merge(dst, src interface{}, opts ...func(*Config)) error {
 	return err
 }
 
-func handleNil(dst reflect.Value) reflect.Value {
-	if !dst.CanSet() {
-		t := reflect.Indirect(reflect.ValueOf(dst)).Type()
-		dsttmp := reflect.New(t).Elem()
-		fmt.Println(dsttmp)
-		initiallize(t, dsttmp)
-		if !isTypeExported(t) {
-			return dsttmp
-		}
-		dsttmp.Set(dst)
-		dst = dsttmp
-	}
-	return dst
-}
-
-func initiallize(t reflect.Type, v reflect.Value) {
-	// fmt.Println("initializeStruct(1.2.1)", t)
-	switch t.Kind() {
-	case reflect.Map:
-		v.Set(reflect.MakeMap(t))
-	case reflect.Slice:
-		v.Set(reflect.MakeSlice(t, 0, 0))
-	case reflect.Chan:
-		v.Set(reflect.MakeChan(t, 0))
-	case reflect.Struct:
-		for i := 0; i < v.NumField(); i++ {
-			f := v.Field(i)
-			ft := t.Field(i)
-			initiallize(ft.Type, f)
-		}
-	case reflect.Ptr:
-		ft := t.Elem()
-		fv := reflect.New(ft)
-		initiallize(ft, fv.Elem())
-		if isTypeExported(ft) {
-			v.Set(fv)
-		}
-	default:
-	}
-}
-
-func handleEmpty(dst, src reflect.Value, overwrite, appendSlice, overwriteWithEmptySrc bool) reflect.Value {
-	dst = handleNil(dst)
-	overwrite = (overwrite && !appendSlice || isEmptyValue(dst))
-	if (!isEmptyValue(src) || overwriteWithEmptySrc) && overwrite {
-		if dst.CanSet() {
-			dst.Set(src)
-		} else {
-			dst = src
-		}
-	}
-	if dst.CanInterface() {
-		dst = reflect.ValueOf(dst.Interface())
-	}
-	return dst
-}
-
 // IsReflectNil is the reflect value provided nil
 func isReflectNil(v reflect.Value) bool {
 	k := v.Kind()

--- a/merge.go
+++ b/merge.go
@@ -101,7 +101,7 @@ func deepMerge(dstIn, src reflect.Value, visited map[uintptr]*visit, depth int, 
 				// copy un-exported struct fields
 				if !isExportedComponent(&structField) {
 					rf := dstCp.Field(i)
-					rf = reflect.NewAt(rf.Type(), unsafe.Pointer(rf.Pointer())).Elem() //nolint:gosec
+					rf = reflect.NewAt(rf.Type(), unsafe.Pointer(rf.UnsafeAddr())).Elem() //nolint:gosec
 					dstRF := dst.Field(i)
 					if !dst.Field(i).CanAddr() {
 						continue

--- a/mergo_test.go
+++ b/mergo_test.go
@@ -345,13 +345,13 @@ func TestEmptyToNotEmptyMaps(t *testing.T) {
 }
 
 func TestMapsWithOverwrite(t *testing.T) {
-	m := map[string]simpleTest{
+	dst := map[string]simpleTest{
 		"a": {},   // overwritten by 16
 		"b": {42}, // overwritten by 0, as map Value is not addressable and it doesn't check for b is set or not set in `n`
 		"c": {13}, // overwritten by 12
 		"d": {61},
 	}
-	n := map[string]simpleTest{
+	src := map[string]simpleTest{
 		"a": {16},
 		"b": {},
 		"c": {12},
@@ -365,12 +365,12 @@ func TestMapsWithOverwrite(t *testing.T) {
 		"e": {14},
 	}
 
-	if err := MergeWithOverwrite(&m, n); err != nil {
+	if err := MergeWithOverwrite(&dst, src); err != nil {
 		t.Fatalf(err.Error())
 	}
 
-	if !reflect.DeepEqual(m, expect) {
-		t.Fatalf("Test failed:\ngot  :\n%#v\n\nwant :\n%#v\n\n", m, expect)
+	if !reflect.DeepEqual(dst, expect) {
+		t.Fatalf("Test failed:\ngot  :\n%#v\n\nwant :\n%#v\n\n", dst, expect)
 	}
 }
 
@@ -885,6 +885,7 @@ func TestBooleanPointer(t *testing.T) {
 }
 
 func TestMergeMapWithInnerSliceOfDifferentType(t *testing.T) {
+<<<<<<< HEAD
 	testCases := []struct {
 		name    string
 		options []func(*Config)
@@ -900,6 +901,13 @@ func TestMergeMapWithInnerSliceOfDifferentType(t *testing.T) {
 			[]func(*Config){WithOverride, WithTypeCheck},
 			"cannot override two slices with different type",
 		},
+=======
+	dst := map[string]interface{}{
+		"foo": []string{"a", "b"},
+	}
+	src := map[string]interface{}{
+		"foo": []int{1, 2},
+>>>>>>> Reduce code complexity
 	}
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
@@ -910,10 +918,15 @@ func TestMergeMapWithInnerSliceOfDifferentType(t *testing.T) {
 				"foo": []int{1, 2},
 			}
 
+<<<<<<< HEAD
 			if err := Merge(&src, &dst, tc.options...); err == nil || !strings.Contains(err.Error(), tc.err) {
 				t.Fatalf("expected %q, got %q", tc.err, err)
 			}
 		})
+=======
+	if err := Merge(&dst, src, WithOverride, WithAppendSlice); err == nil {
+		t.Fatal("expected an error, got nothing")
+>>>>>>> Reduce code complexity
 	}
 }
 

--- a/mergo_test.go
+++ b/mergo_test.go
@@ -129,10 +129,10 @@ func TestComplexStruct(t *testing.T) {
 }
 
 func TestComplexStructWithOverwrite(t *testing.T) {
-	a := complexTest{simpleTest{1}, 1, "do-not-overwrite-with-empty-value"}
-	b := complexTest{simpleTest{42}, 2, ""}
+	a := complexTest{St: simpleTest{1}, sz: 1, ID: "do-not-overwrite-with-empty-value"}
+	b := complexTest{St: simpleTest{42}, sz: 2, ID: ""}
 
-	expect := complexTest{simpleTest{42}, 1, "do-not-overwrite-with-empty-value"}
+	expect := complexTest{St: simpleTest{42}, sz: 1, ID: "do-not-overwrite-with-empty-value"}
 	if err := MergeWithOverwrite(&a, b); err != nil {
 		t.FailNow()
 	}
@@ -156,7 +156,7 @@ func TestPointerStruct(t *testing.T) {
 }
 
 type embeddingStruct struct {
-	embeddedStruct
+	A embeddedStruct
 }
 
 type embeddedStruct struct {
@@ -359,7 +359,7 @@ func TestMapsWithOverwrite(t *testing.T) {
 	}
 	expect := map[string]simpleTest{
 		"a": {16},
-		"b": {},
+		"b": {42},
 		"c": {12},
 		"d": {61},
 		"e": {14},
@@ -536,23 +536,13 @@ func TestMergeUsingStructAndMap(t *testing.T) {
 }
 func TestMaps(t *testing.T) {
 	m := map[string]simpleTest{
-		"a": {},
-		"b": {42},
-		"c": {13},
-		"d": {61},
+		"a": {0}, "b": {42}, "c": {13}, "d": {61},
 	}
 	n := map[string]simpleTest{
-		"a": {16},
-		"b": {},
-		"c": {12},
-		"e": {14},
+		"a": {16}, "b": {}, "c": {12}, "e": {14},
 	}
 	expect := map[string]simpleTest{
-		"a": {0},
-		"b": {42},
-		"c": {13},
-		"d": {61},
-		"e": {14},
+		"a": {16}, "b": {42}, "c": {13}, "d": {61}, "e": {14},
 	}
 
 	if err := Merge(&m, n); err != nil {
@@ -562,7 +552,7 @@ func TestMaps(t *testing.T) {
 	if !reflect.DeepEqual(m, expect) {
 		t.Fatalf("Test failed:\ngot  :\n%#v\n\nwant :\n%#v\n\n", m, expect)
 	}
-	if m["a"].Value != 0 {
+	if m["a"].Value != 16 {
 		t.Fatalf(`n merged in m because I solved non-addressable map values TODO: m["a"].Value(%d) != n["a"].Value(%d)`, m["a"].Value, n["a"].Value)
 	}
 	if m["b"].Value != 42 {

--- a/mergo_test.go
+++ b/mergo_test.go
@@ -893,12 +893,12 @@ func TestMergeMapWithInnerSliceOfDifferentType(t *testing.T) {
 		{
 			"With override and append slice",
 			[]func(*Config){WithOverride, WithAppendSlice},
-			"cannot append two slices with different type",
+			"cannot append two different types (slice, slice)",
 		},
 		{
 			"With override and type check",
 			[]func(*Config){WithOverride, WithTypeCheck},
-			"cannot override two slices with different type",
+			"cannot append two different types (slice, slice)",
 		},
 	}
 	for _, tc := range testCases {

--- a/mergo_test.go
+++ b/mergo_test.go
@@ -885,7 +885,6 @@ func TestBooleanPointer(t *testing.T) {
 }
 
 func TestMergeMapWithInnerSliceOfDifferentType(t *testing.T) {
-<<<<<<< HEAD
 	testCases := []struct {
 		name    string
 		options []func(*Config)
@@ -901,13 +900,6 @@ func TestMergeMapWithInnerSliceOfDifferentType(t *testing.T) {
 			[]func(*Config){WithOverride, WithTypeCheck},
 			"cannot override two slices with different type",
 		},
-=======
-	dst := map[string]interface{}{
-		"foo": []string{"a", "b"},
-	}
-	src := map[string]interface{}{
-		"foo": []int{1, 2},
->>>>>>> Reduce code complexity
 	}
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
@@ -918,15 +910,10 @@ func TestMergeMapWithInnerSliceOfDifferentType(t *testing.T) {
 				"foo": []int{1, 2},
 			}
 
-<<<<<<< HEAD
 			if err := Merge(&src, &dst, tc.options...); err == nil || !strings.Contains(err.Error(), tc.err) {
 				t.Fatalf("expected %q, got %q", tc.err, err)
 			}
 		})
-=======
-	if err := Merge(&dst, src, WithOverride, WithAppendSlice); err == nil {
-		t.Fatal("expected an error, got nothing")
->>>>>>> Reduce code complexity
 	}
 }
 

